File: backend/internal/controllers/errors_test.go
package controllers

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"

	"nexia-backend/internal/services"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// TestRespondWithServiceErrorMapping verifies every branch of the switch —
// each sentinel service error maps to the expected HTTP status and error code.
// Previously only partially covered through integration tests; this pins the
// contract at the unit level.
func TestRespondWithServiceErrorMapping(t *testing.T) {
	gin.SetMode(gin.TestMode)

	cases := []struct {
		name     string
		err      error
		wantCode int
		wantKey  string
	}{
		{"account not found", services.ErrAccountNotFound, http.StatusUnauthorized, "ACCOUNT_NOT_FOUND"},
		{"unauthorized", services.ErrUnauthorized, http.StatusUnauthorized, "UNAUTHORIZED"},
		{"validation wrapped", fmt.Errorf("%w: bad input", services.ErrValidation), http.StatusBadRequest, "VALIDATION_ERROR"},
		{"sentinel not found", services.ErrNotFound, http.StatusNotFound, "NOT_FOUND"},
		{"gorm record not found", gorm.ErrRecordNotFound, http.StatusNotFound, "NOT_FOUND"},
		{"ai unavailable", services.ErrAIUnavailable, http.StatusServiceUnavailable, "AI_UNAVAILABLE"},
		{"email not verified", services.ErrEmailNotVerified, http.StatusForbidden, "EMAIL_NOT_VERIFIED"},
		{"email conflict", services.ErrEmailConflict, http.StatusConflict, "EMAIL_CONFLICT"},
		{"unknown error falls through to 500", errors.New("some wild error"), http.StatusInternalServerError, "SERVER_ERROR"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			w := httptest.NewRecorder()
			c, _ := gin.CreateTestContext(w)
			respondWithServiceError(c, tc.err)

			if w.Code != tc.wantCode {
				t.Fatalf("status: want %d got %d", tc.wantCode, w.Code)
			}
			var body map[string]any
			if err := json.Unmarshal(w.Body.Bytes(), &body); err != nil {
				t.Fatalf("unmarshal: %v", err)
			}
			errObj, ok := body["error"].(map[string]any)
			if !ok {
				t.Fatalf("expected error envelope, got %s", w.Body.String())
			}
			if errObj["code"] != tc.wantKey {
				t.Fatalf("code: want %s got %v", tc.wantKey, errObj["code"])
			}
		})
	}
}

File: apps/api/src/services/errors.test.ts
import { describe, test, expect } from "bun:test";
import {
  ServiceError,
  ErrorKind,
  isServiceError,
  errValidation,
  errNotFound,
  errUnauthorized,
  errAccountNotFound,
  errAIUnavailable,
  errEmailNotVerified,
  errEmailConflict,
} from "../services/errors";
import { respondWithServiceError, respondError } from "../utils/http";

describe("service errors", () => {
  test("error kind constants", () => {
    expect(ErrorKind.NotFound).toBe("not_found");
    expect(ErrorKind.Unauthorized).toBe("unauthorized");
    expect(ErrorKind.Validation).toBe("validation");
  });

  test("ServiceError constructor", () => {
    const err = new ServiceError(ErrorKind.NotFound);
    expect(err.kind).toBe("not_found");
    expect(err.message).toBe("not_found");
    expect(err.name).toBe("ServiceError");
  });

  test("ServiceError with custom message", () => {
    const err = new ServiceError(ErrorKind.Validation, "Bad input");
    expect(err.message).toBe("Bad input");
  });

  test("isServiceError type guard", () => {
    expect(isServiceError(new ServiceError(ErrorKind.NotFound))).toBe(true);
    expect(isServiceError(new Error("regular"))).toBe(false);
    expect(isServiceError("string")).toBe(false);
  });

  test("errValidation helper", () => {
    const err = errValidation("Too many songs");
    expect(err.kind).toBe(ErrorKind.Validation);
    expect(err.message).toBe("Too many songs");
  });

  test("errNotFound helper", () => {
    const err = errNotFound();
    expect(err.kind).toBe(ErrorKind.NotFound);
  });

  test("errUnauthorized helper", () => {
    const err = errUnauthorized();
    expect(err.kind).toBe(ErrorKind.Unauthorized);
  });
});

// respondWithServiceError/respondError now use new Response() directly
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const c = undefined as any;

async function readBody(resp: Response): Promise<{ error: { code: string; message: string } }> {
  return resp.json() as Promise<{ error: { code: string; message: string } }>;
}

describe("respondWithServiceError", () => {
  const cases = [
    { err: errAccountNotFound(), status: 401, code: "ACCOUNT_NOT_FOUND" },
    { err: errUnauthorized(), status: 401, code: "UNAUTHORIZED" },
    { err: errValidation("bad request"), status: 400, code: "VALIDATION_ERROR" },
    { err: errNotFound(), status: 404, code: "NOT_FOUND" },
    { err: errAIUnavailable(), status: 503, code: "AI_UNAVAILABLE" },
    { err: errEmailNotVerified(), status: 403, code: "EMAIL_NOT_VERIFIED" },
    { err: errEmailConflict(), status: 409, code: "EMAIL_CONFLICT" },
    { err: new Error("something broke"), status: 500, code: "SERVER_ERROR" },
  ];

  for (const tc of cases) {
    test(`${tc.code} → ${tc.status}`, async () => {
      const resp = respondWithServiceError(c, tc.err);
      expect(resp.status).toBe(tc.status);
      const body = await readBody(resp);
      expect(body.error.code).toBe(tc.code);
      expect(typeof body.error.message).toBe("string");
    });
  }
});

describe("respondError", () => {
  test("writes structured error response", async () => {
    const resp = respondError(c, 400, "BAD_REQUEST", "Something wrong");
    expect(resp.status).toBe(400);
    const body = await readBody(resp);
    expect(body.error.code).toBe("BAD_REQUEST");
    expect(body.error.message).toBe("Something wrong");
  });
});
